networking/networking.go

```diff
@@ -82,69 +61,115 @@ func Setup(rktRoot string, podID types.UUID, fps []ForwardedPort) (*Networking,
 		},
 	}
 
-	defer func() {
-		// cleanup on error
-		if err != nil {
-			n.Teardown()
-		}
-	}()
-
-	if n.hostNS, n.podNS, err = basicNetNS(); err != nil {
+	hostNS, podNS, err := basicNetNS()
+	if err != nil {
 		return nil, err
 	}
 	// we're in podNS!
+	n.hostNS = hostNS
+
+	nspath := n.podNSPath()
 
-	n.podNSPath = filepath.Join(rktRoot, "netns")
-	if err = bindMountFile(selfNetNS, n.podNSPath); err != nil {
+	if err = bindMountFile(selfNetNS, nspath); err != nil {
 		return nil, err
 	}
 
-	nets, err := n.loadNets()
+	defer func() {
+		if err != nil {
+			if err := syscall.Unmount(nspath, 0); err != nil {
+				log.Printf("Error unmounting %q: %v", nspath, err)
+			}
+		}
+	}()
+
+	n.nets, err = n.loadNets()
 	if err != nil {
 		return nil, fmt.Errorf("error loading network definitions: %v", err)
 	}
 
-	err = withNetNS(n.podNS, n.hostNS, func() error {
-		n.nets, err = n.setupNets(n.podNSPath, nets)
-		if err != nil {
+	err = withNetNS(podNS, hostNS, func() error {
+		if err := n.setupNets(n.nets); err != nil {
 			return err
 		}
+		return n.forwardPorts(fps, n.GetDefaultIP())
+	})
+	if err != nil {
+		return nil, err
+	}
 
-		if len(n.nets) == 0 {
-			return fmt.Errorf("no nets successfully configured")
-		}
+	return &n, nil
+}
 
-		// last net is the default
-		n.MetadataIP = n.nets[len(n.nets)-1].ip
-		n.HostIP = n.nets[len(n.nets)-1].hostIP
+// Loadcreates the Networking object from saved state.
+// Assumes the current netns is that of the host.
+func Load(rktRoot string, podID *types.UUID) (*Networking, error) {
+	// the current directory is pod root
+	pdirfd, err := syscall.Open(".", syscall.O_RDONLY|syscall.O_DIRECTORY, 0)
```

Comment:
hmm, surely either use rktRoot here or omit it entirely as an arg? or can you use it for testing?